### PR TITLE
added colored_search and some examples

### PR DESCRIPTION
**Pull request by davide710**
Hi, I think this could close [#40](url) and [#16:](url) I created the function "colored_search", that calls `cv2.matchTemplate` and then checks the color distributions of the image and of the found region on the screen to see if they are compatible. This way you can eliminate the false positives when you want to search for an image and you want the program to find it only if it appears in that specific colors.
I also added the file "test_colored_search.py" and some images to provide a basic demonstration of the mistakes made by `cv2.matchTemplate` which are corrected with my function: as you can see, the former detects images of a red, green or yellow circles in a traffic light even when the corresponding light is off.
I would be happy to contribute to this repository, @drov0 please let me know what you think about this and if the PR needs some changes.
Davide

### CODE REVIEW

**Comment by davide710** [comment]
Hello @drov0, @DashBlacK, @BluBambu, sorry to bother you, is this repository still active? I'm still doing my first contributions so I may have done something wrong, did you at least see the pull request? Did you get a notification or something? Thank you in advance. Davide

**Comment by drov0** [comment]
Hi @davide710 The repository is still active, I saw your merge request thank you :) I will review it when I get a bit of time

File: test_colored_search.py [hunk 1]
```diff
@@ -0,0 +1,147 @@
+import cv2
+import numpy as np
+
+
+def chi_2(l1, l2):
```
Comment (drov0) [inline test_colored_search.py]: Directly import the chi_2 from the main imagesearch file please

File: python_imagesearch/imagesearch.py [hunk 2]
```diff
@@ -327,3 +327,77 @@ def imagesearch_click(image, action, delay, offset=5, precision=0.8):
...
+Returns the coordinates of the given image but only if the colors match
+
+input :
+image : path to the image file (see opencv imread for supported types)
+precision : the higher, the lesser tolerant and fewer false positives are found default is 0.8
+
+returns :
+the top left corner coordinates of the element if found as an array [x,y] or [-1,-1] if not
+"""
+def colored_search(image, precision=0.8):
```
Comment (drov0) [inline python_imagesearch/imagesearch.py]: call it imagesearch_colored instead please to match the rest of the repo

File: test_colored_search.py [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+    chi = 0
+    for i in range(len(l2)):
+        if l2[i] != 0:
+            chi += (l1[i] - l2[i])**2 / l2[i]
+        else:
+            chi += l1[i]**2
+
+    return chi / len(l2)
+
+def f(base, template):
```
Comment (drov0) [inline test_colored_search.py]: Given that those are unit tests (as opposed to say simple_example.py) , I think it would make sense to move this file and the "images" directory into a  "tests" directory.

File: test_colored_search.py [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+    base = np.array(base)
```
Comment (drov0) [inline test_colored_search.py]: This project isn't aimed to follow the absolute best practices so I don't care too much it's more of a side project lib for me to help me on my projects really, so none of this is like this I expect you to do to get this merged.

This is more something to keep in mind for future contributions to this repository or others and for yourself.
While these tests are useful for me to see the inner workings and "aha this works". they add little value to the overall project because they don't test the "production" function (imagesearch_colored or chi_2) everything is copy pasted, which means that if something breaks in the real function, those tests would continue to work, kind of defeating the purpose of those unit tests.

So if I were to do things the correct way to be unit_testable I would split colored_imagesearch into two functions, one that captures the screen and another that does the computations, that way you can execute unit tests on the actual matching logic using pre-made screenshots (like you're doing now) by directly importing them from the "main file".

Another point is that it's a little difficult to know if tests are actually successful or not, "is this supposed to return -1 or is this a negative test ?". 
A first step could be to add asserts  https://www.w3schools.com/python/ref_keyword_assert.asp which doesn't add a lot of weight to the overall test. And then if we wanted to "professionalize" even more we could add proper test cases with pytest etc.

Again I don't require you to do that, though, but feel free if you want to.

**Review by drov0 — CHANGES_REQUESTED** [review]

File: test_colored_search.py (hunk 4 above)
Comment (davide710) [reply test_colored_search.py]: Thank you for the advice (I am absolutely not an expert).
I added test_colored_search.py and the images because I wanted to give you the quickest and easiest demonstration of what improvement I was bringing, exactly so that you could say
> aha this works

Comment (drov0) [reply test_colored_search.py]: Haha, I did test locally with other images just in case though

**Comment by davide710** [comment]
Hey @drov0 , I removed those "tests", changed the function name, and splitted it in two functions as you suggested.

**Comment by drov0** [comment]
LGTM, Thank you !
